src/org/exist/xquery/functions/transform/Transform.java

```diff
@@ -360,71 +348,75 @@ public Sequence eval(Sequence[] args, Sequence contextSequence) throws XPathExce
+
+                        stylesheet = TemplatesFactory.stylesheet(uri, context.getModuleLoadPath(), attributes, useCache);
                     }
                 }
-                if (templates == null) {
-                    if (stylesheetItem instanceof Document) {
-                        String uri = ((Document) stylesheetItem).getDocumentURI();
-
-						/*
-                         * This must be checked because in the event the stylesheet is
-						 * an in-memory document, it will cause an NPE
-						 */
-                        if (uri != null) {
-                            uri = uri.substring(0, uri.lastIndexOf('/'));
-                            factory.setURIResolver(new EXistURIResolver(context.getBroker(), uri));
-                        }
-                    }
-                    templates = getSource(factory, stylesheetNode);
+                if (stylesheet == null) {
+                    stylesheet = TemplatesFactory.stylesheet(
+                        getContext().getBroker(),
+                        stylesheetNode,
+                        context.getModuleLoadPath()
+                    );
                 }
             } else {
+                String baseUri = context.getModuleLoadPath();
                 if (stylesheetItem instanceof Document) {
-                    String uri = ((Document) stylesheetItem).getDocumentURI();
-
-					/*
-					 * This must be checked because in the event the stylesheet is 
-					 * an in-memory document, it will cause an NPE
-					 */
-                    if (uri != null) {
-                        uri = uri.substring(0, uri.lastIndexOf('/'));
-                        factory.setURIResolver(new EXistURIResolver(context.getBroker(), uri));
+                    baseUri = ((Document) stylesheetItem).getDocumentURI();
+
+                    /*
+                     * This must be checked because in the event the stylesheet is
+                     * an in-memory document, it will cause an NPE
+                     */
+                    if (baseUri == null) {
+                        baseUri = context.getModuleLoadPath();
+                    } else {
+                        baseUri = baseUri.substring(0, baseUri.lastIndexOf('/'));
                     }
                 }
 
-                final String stylesheet = stylesheetItem.getStringValue();
-                templates = getSource(factory, stylesheet);
+                final String uri = stylesheetItem.getStringValue();
+
+                stylesheet = TemplatesFactory.stylesheet(uri, baseUri, attributes, useCache);
             }
-            handler = factory.newTransformerHandler(templates);
+
+            handler = stylesheet.newTransformerHandler(getContext().getBroker(), errorListener);
 
             if (options != null) {
                 setParameters(options, handler.getTransformer());
             }
 
-        } catch (final TransformerConfigurationException e) {
+        } catch (final Exception e) {
```

Comment:
Because it does not show what exceptions are under consideration. Also it will catch RuntimeException's which it should not.